deploy/utils.go

```
@@ -0,0 +1,168 @@
+// SudoCopyFile copies the file from the source to dest as root
+func SudoCopyFile(dest, src string) error {
+	cmd := exec.Command("sudo", "cp", src, dest)
+	if err := cmd.Run(); err != nil {
+		return errors.Wrapf(err, "Error running: %v", cmd.Args)
+	}
+	return nil
+}
+
+// SudoMakeDirectory makes the desired set of directories as root
+func SudoMakeDirectory(path string) error {
+	cmd := exec.Command("sudo", "mkdir", "-p", path)
+	if err := cmd.Run(); err != nil {
+		return errors.Wrapf(err, "Error running: %v", cmd.Args)
+	}
+	return nil
+}
+
+var systemdServiceData = `
+[Unit]
+Description=%s service
+After=network.target
+
+[Service]
+Type=simple
+ExecStart=/usr/local/bin/%s --cacert=%s --cert=%s --v 3
+Restart=no
+KillMode=process
+
+[Install]
+WantedBy=multi-user.target
+`
+
+// DownloadTool downloads the desired tool from the server
+func DownloadTool(deployServerHost, tool string) (string, error) {
+	url := fmt.Sprintf("%s/download/%s", deployServerHost, tool)
+	fmt.Printf("Downloading %s: %s\n", tool, url)
+	r, err := http.Get(url)
+
+	if err != nil {
+		return "", errors.Wrap(err, "Error downloading tool")
+	}
+	defer r.Body.Close()
+
+	if r.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("Expected OK when downloading binary: %s", r.Status)
+	}
+
+	f, err := ioutil.TempFile("", tool)
+	if err != nil {
+		return "", errors.Wrap(err, "Error creating temporary file for tool")
+	}
+	defer f.Close()
+
+	if _, err := io.Copy(f, r.Body); err != nil {
+		return "", errors.Wrap(err, "Error downloading file")
+	}
+
+	if err := os.Chmod(f.Name(), 0755); err != nil {
+		return "", errors.Wrap(err, "Error changing file permission")
+	}
+
+	return f.Name(), nil
+}
+
+// InstallTool installs a tool to its final destination and manages it via systemd
+func InstallTool(toolPath, tool, caCertPath, certPath string) error {
+	fmt.Printf("Installing %s\n", tool)
+
+	fmt.Printf("Stopping %s\n", tool)
+	cmd := exec.Command("sudo", "systemctl", "stop", tool)
+	cmd.Run()
```

We're ignoring an error here.